### PR TITLE
Adds shutdown procedure and logic to HWP PMX agent

### PR DESCRIPTION
**Pull request by jlashner**
This PR implements a shutdown procedure for the HWP PMX agent.

## Description

This shows how one may use the hwp-supervisor session data to trigger a shutdown procedure, and how a shutdown might lock out specific tasks or processes until it is reversed.

This implements three additional operations:
- monitor_shutdown is a process that continuously monitors hwp-supervisor rotation action to determine whether or not a shutdown should be triggered. This will trigger a shutdown if either the action is 'stop', or it has been `no_data_timeout` since the last 'ok' action received. The action will be published to grafana, so an alarm can be set that triggers if the action is not 'ok' for a certain amount of time.
- The 'initiate_shutdown' task will set `shutdown_mode=True`, and will tell the PMX to cut power and to ignore external voltage from the PID.
- The `reverse_shutdown` task sets `shutdown_mode=False`

While `shutdown_mode` is True, any task that tries to change the PMX state in a way that might interfere with the shutdown (such as `set_on`, `set_i`, `set_v`...) will fail.

## Motivation and Context

Required for emergency HWP shutdown

## How Has This Been Tested?

This has not been tested yet

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: socs/agents/hwp_pmx/agent.py [hunk 1]
```diff
@@ -250,6 +285,83 @@ def _stop_acq(self, session, params=None):
...
+        self.log.warn("INITIATING SHUTDOWN")
+
+        with self.lock.acquire_timeout(10, job='shutdown') as acquired:
+            if not acquired:
+                self.log.error("Could not acquire lock for shutdown.")
+                return False, "Could not acquire lock."
+
+            self.shutdown_mode = True
+            self.dev.ign_external_voltage()
+            self.dev.stop()
```
Comment (bbixler500) [inline socs/agents/hwp_pmx/agent.py:301]: I don't see a "stop()" method for self.dev as defined in "/socs/agents/hwp_pmx/drivers/PMX_ethernet.py". If you're intending to turn off the PMX output I would use "self.dev.turn_off()" instead.

Comment (jlashner) [reply socs/agents/hwp_pmx/agent.py:301]: Ah great, thanks for catching this!

File: socs/agents/hwp_pmx/agent.py [hunk 2]
```diff
@@ -250,6 +285,83 @@ def _stop_acq(self, session, params=None):
...
+                if not self.shutdown_mode:
+                    self.agent.start('initiate_shutdown')
+
+            data = {
+                'data': {'rotation_action': action},
+                'block_name': 'rotation_action',
+                'timestamp': time.time()
+            }
+
+            self.agent.publish_to_feed('gripper_action', data)
```
Comment (bbixler500) [inline socs/agents/hwp_pmx/agent.py:351]: Should this be 'rotation_action' instead of 'gripper_action'?

File: socs/agents/hwp_pmx/agent.py [hunk 3]
```diff
@@ -20,9 +21,13 @@ class HWPPMXAgent:
         ip (str): IP address for the PMX Kikusui power supply
         port (str): Port for the PMX Kikusui power supply
         f_sample (float): sampling frequency (Hz)
+        supervisor_id (str): ID of HWP supervisor
+        no_data_timeout (float): Time (in seconds) to wait between receiving
+            'no_data' actions from the supervisro and triggering a shutdown
```
Comment (ykyohei) [inline socs/agents/hwp_pmx/agent.py:26]: typo 'supervisro'

**Review by ykyohei — APPROVED** [review]
Sorry for late review, and thank you for your work.
Shutdown procedure looks reasonable to me.

**Comment by BrianJKoopman** [comment]
Sorry, I had to rebase the branch this was being PR'd onto (in https://github.com/simonsobs/socs/pull/421). This has messed up this branch a bit, as it still has all of those rebased commits. You should probably rebase on `main`.

File: socs/agents/hwp_pmx/agent.py [hunk 4]
```diff
@@ -260,6 +295,83 @@ def _stop_acq(self, session, params=None):
...
+            # trigger shutdown
+            elif action == 'no_data':
+                if (time.time() - last_ok_time) > self.no_data_timeout:
+                    if not self.shutdown_mode:
+                        self.agent.start('initiate_shutdown')
+
+            # If action is 'shutdown', trigger shutdown
+            elif action == 'stop':
+                if not self.shutdown_mode:
+                    self.agent.start('initiate_shutdown')
```
Comment (BrianJKoopman) [inline socs/agents/hwp_pmx/agent.py:353]: You should also pass `params` here, even if just `None`. (If `initiate_shutdown` had the params decorator, this as is would run into an error.) Same goes for other direct calls to `.start()`.

File: socs/agents/hwp_pmx/agent.py [hunk 5]
```diff
@@ -260,6 +295,83 @@ def _stop_acq(self, session, params=None):
         else:
             return False, 'acq is not currently running'
 
+    def initiate_shutdown(self, session, params):
+        """
```
Comment (BrianJKoopman) [inline socs/agents/hwp_pmx/agent.py:299]: These task docstrings all need the method signature override at the top, and to be marked with `**Task**` or `**Process**` appropriately. See the [example here](https://ocs.readthedocs.io/en/main/developer/agent_references/documentation.html#task-and-process-documentation).

**Review by BrianJKoopman — CHANGES_REQUESTED** [review]
A few small ocs related requests. Also, after seeing the import from the supervisor agent, I do think we should merge this and #427 together before merging to `main`.

**Review by BrianJKoopman — APPROVED** [review]
Looks good to merge into [hwp_supervisor](https://github.com/simonsobs/socs/tree/hwp_supervisor).

**Pull request merged**